Bring CDEPS Inline capability

uturuncoglu opened the pull request:
## Description

This PR aims to bring CDEPS Inline capability to FV3 to update desired fields.

### Issue(s) addressed

N/A

## Testing

How were these changes tested?  
What compilers / HPCs was it tested with?  
Are the changes covered by regression tests? (If not, why? Do new tests need to be added?)  
Have the ufs-weather-model regression test been run? On what platform?  
- Will the code updates change regression test baseline? If yes, why? Please show the baseline directory below.
- Please commit the regression test log files in your ufs-weather-model branch

We have a limited test with SRW HRRR configuration to update SST and other sea ice related variables over Great Lakes.

## Dependencies

If testing this branch requires non-default branches in other repositories, list them.
Those branches should have matching names (ideally)

Do PRs in upstream repositories need to be merged first?
If so add the "waiting for other repos" label and list the upstream PRs
- waiting on https://github.com/ufs-community/ccpp-physics/pull/298

uturuncoglu committed:
initial commit for cdeps inline capability

uturuncoglu committed:
more work for cdeps inline

uturuncoglu committed:
clean code

uturuncoglu committed:
Merge remote-tracking branch 'upstream/develop' into feature/inline

uturuncoglu committed:
sync model

uturuncoglu committed:
sync ccpp

uturuncoglu committed:
sync submodules

uturuncoglu committed:
update inline capability

uturuncoglu committed:
more work for cdeps inline

uturuncoglu committed:
more work for cdeps inline and ccpp connection

uturuncoglu committed:
update implementation

uturuncoglu committed:
minor fix

uturuncoglu committed:
make cdeps inline runtime configurable

Commit:
more work for inline

uturuncoglu committed:
sync ccpp

uturuncoglu committed:
sync fv3

uturuncoglu committed:
update pointer to branch

uturuncoglu committed:
sync fv3

uturuncoglu committed:
update ccpp physics

uturuncoglu committed:
more work for cdeps inline

uturuncoglu committed:
sync fv3

uturuncoglu committed:
fix issues related to the sync

uturuncoglu commented:
Note that the implementation is not working after - https://github.com/NOAA-EMC/fv3atm/commit/aa1484378de6cc665910b63a691a1475809f6bb3. The data provided by CDEPS can not be passed correctly. Probably the approach used in `fv3/atmos_model.F90` needs to be fixed.

DusanJovic-NOAA commented:
@uturuncoglu Can you check the lower/upper bounds of datar82d array in setup_inlinedata:

This loop:

```
          do j = jsc, jec
             do i = isc, iec
                nb = Atm_block%blkno(i,j)
                ix = Atm_block%ixp(i,j)
                im = GFS_control%chunk_begin(nb)+ix-1
                GFS_Coupling%fice_dat(im) = datar82d(i,j)
             end do
          end do
``` 
assumes the indexes of datar82d are global. But when the ESMF_Field holding this pointer is created:

https://github.com/NOAA-EMC/fv3atm/blob/3145f5b45e244a0949b7082d8bb3057df269b619/cpl/module_inline.F90#L155

`indexflag=ESMF_INDEX_DELOCAL` is used. Try to use ESMF_INDEX_GLOBAL.

uturuncoglu commented:
@DusanJovic-NOAA Let me try. Thanks.

uturuncoglu commented:
@DusanJovic-NOAA JFYI, I am passing grid component to CDEPS inline with following call

```
call stream_init(fcstGridComp(cpl_grid_id), clock, rc)
```

and grid for `fcstGridComp` is defined like following in the `FV3/fv3/module_fcst_grid_comp.F90`.

```
      grid = ESMF_GridCreateCubedSphere(tileSize=tilesize, &
                                        coordSys=ESMF_COORDSYS_SPH_RAD, & 
                                        coordTypeKind=grid_typekind, &
                                        regDecompPTile=decomptile, &
                                        decompflagPTile=decompflagPTile, &
                                        name="fcst_grid", rc=rc)
```

As I know that the default value for the index flag is `ESMF_INDEX_DELOCAL` in here. So, if I set `ESMF_INDEX_GLOBAL`, I am getting following error,

```
20250711 155429.969 ERROR            PET158 ESMF_FieldEmpty.F90:19545 ESMF_FieldEmptyCompGB Argument sizes do not match  - - user specified indexflag must be identical with Grid indexflag
20250711 155429.970 ERROR            PET158 ESMF_FieldCreate.F90:3232 ESMF_FieldCreateGBData Argument sizes do not match  - Internal subroutine call returned Error
20250711 155429.970 ERROR            PET158 ESMF_FieldCreate.F90:9567 ESMF_FieldCreateGridData Argument sizes do not match  - Internal subroutine call returned Error
```

The fields needs to be created with `ESMF_INDEX_DELOCAL` to be consistent with grid definition.

uturuncoglu commented:
The `datar82d` indexes are local in my case. So, I updated `datar82d(i,j)` to `datar82d(i-isc+1,j-jsc+1)` to make the indexes also local. Then, run my configuration again but did not helped too much. 

This is what CDEPS provides in terms of mask - stored in `datar82d`

<img width="795" height="485" alt="Screenshot 2025-07-11 at 4 32 42 PM" src="https://github.com/user-attachments/assets/6f6a7a3f-907b-43d5-86a8-afb23bb959a5" />

and following is from FV3 output file `phyf000.nc` which is in CCPP side (using CCPP debugging I added it to the FV3 history file),

<img width="788" height="476" alt="Screenshot 2025-07-11 at 4 32 26 PM" src="https://github.com/user-attachments/assets/373de32b-7c1b-4ef7-82cf-e33b87113be0" />

It seems that block positions are fine but maybe there is an issue with `im = GFS_control%chunk_begin(nb)+ix-1` calculation. Any idea?

uturuncoglu commented:
@DusanJovic-NOAA I also think that the data in CCPP side seems rotated or reversed in one direction. Not sure.

DeniseWorthen commented:
@uturuncoglu I'm going to guess that you've tried this in debug mode?

DusanJovic-NOAA commented:
https://github.com/NOAA-EMC/fv3atm/blob/3145f5b45e244a0949b7082d8bb3057df269b619/ccpp/data/GFS_typedefs.meta#L3427

Try to change dimensions attribute here to `horizontal_dimension`.

uturuncoglu commented:
@DeniseWorthen Unfortunately, no. Let me try, it could catch some error in the implementation.

uturuncoglu commented:
@DusanJovic-NOAA I also check SFC_Ocean and Ice and it is defined like `horizontal_loop_extent`, let me know if you still want me to replace all `horizontal_loop_extent` with `horizontal_dimension`.

DusanJovic-NOAA commented:
Other members of the `Coupling` DDT like `hflx_lnd` use `horizontal_dimension`. I think you should try. And definitely compile the model with DEBUG=ON

uturuncoglu commented:
@DusanJovic-NOAA Okay. Let me try.

uturuncoglu commented:
@DusanJovic-NOAA @DeniseWorthen In debug mode I am getting error like following,

```
forrtl: error (72): floating overflow
Image              PC                Routine            Line        Source
libpthread-2.31.s  00001494815648C0  Unknown               Unknown  Unknown
fv3.exe            00000000014B5A69  fv3atm_history_io         528  fv3atm_history_io.F90
libiomp5.so        000014948129CC43  __kmp_invoke_micr     Unknown  Unknown
libiomp5.so        0000149481224FD1  Unknown               Unknown  Unknown
libiomp5.so        0000149481223716  __kmp_fork_call       Unknown  Unknown
libiomp5.so        00001494811E5C88  __kmpc_fork_call      Unknown  Unknown
fv3.exe            00000000014A7F9D  fv3atm_history_io         525  fv3atm_history_io.F90
fv3.exe            00000000014A26EB  fv3atm_history_io         464  fv3atm_history_io.F90
fv3.exe            000000000147908F  fv3atm_history_io         131  fv3atm_history_io.F90
fv3.exe            0000000001A3FCB7  atmos_model_mod_m        1032  atmos_model.F90
fv3.exe            000000000144BB58  module_fcst_grid_        1441  module_fcst_grid_comp.F90
```

This might be related with having extra variables in CCPP side. Anyway, I'll look at it more closely. Also, I did not change the meta file yet.

uturuncoglu commented:
@DusanJovic-NOAA I am trying to change the dimension to `horizontal_dimension` but getting following error from CCPP. Any idea?

```
Exception: Legacy extension DISABLED: replacing variable 'horizontal_dimension' with 'horizontal_loop_extent' in table sfc_data_run
```

DusanJovic-NOAA commented:
@uturuncoglu Maybe you should leave the dimension name in `sfc_data.meta` to be `horizontal_loop_extent`.

See, for example, how the dimension name of `hflx_lnd` is horizonatal_dimension in GFS_typedefs.meta:

https://github.com/NOAA-EMC/fv3atm/blob/3145f5b45e244a0949b7082d8bb3057df269b619/ccpp/data/GFS_typedefs.meta#L2992-L2999

but horizontal_loop_extent in sfc_land.meta:

https://github.com/ufs-community/ccpp-physics/blob/c86be50a34c600423fcd5df558ff802d5286af10/physics/SFC_Models/Land/sfc_land.meta#L246-L254

I do not fully understand how this meta specifications are supposed to work, it's all very confusing to me.

uturuncoglu commented:
@DusanJovic-NOAA Thanks for your help. I think the issue solved after I used `horizontal_dimension` in GFS_data file but not in CCPP side. I am doing more test now to confirm. I'll update you when I have more information. @DeniseWorthen At this point, I am running in debug mode without any issue.

uturuncoglu committed:
fix for cdeps inline

uturuncoglu committed:
set debug level to 0 and update physics

uturuncoglu commented:
@DusanJovic-NOAA @DeniseWorthen I did some test and it seems the issue is fixed. Here are some plots/animations that compares two run (with and without inline implementation) for surface temperature and latent heat.

**2m temperature:**
![plotf_tmpsfc_2d_v2](https://github.com/user-attachments/assets/3d666bde-bef8-4b82-8dd7-5f41d35de63a)

****Latent Haet flux:**
![plot_lhtfl_2d_v2](https://github.com/user-attachments/assets/97438a09-24be-4324-9c8f-1a07c31f72da)

DeniseWorthen commented:
@uturuncoglu Regarding an RT, what exactly would be the configuration? I think I remember you mentioning something about the FV ocean model this was using (FVcom?). Would it require adding that as a new component?

uturuncoglu commented:
@DeniseWorthen The test that aim currently using is COUNS domain and standalone atmosphere. It uses FVCOM model output to update Great Lakes surface conditions (temperature and ice variables). So, there is no any active ocean component.  CDEPS Inline just reads the data and pass it to the CCPP.

DusanJovic-NOAA commented on ccpp/data/GFS_typedefs.F90:
```diff
@@ -792,6 +799,9 @@ module GFS_typedefs
     logical              :: use_med_flux    !< default .false. - i.e. don't use atmosphere-ocean fluxes imported from mediator
 
+!--- cdeps inline parameters
+    logical              :: use_inline      !< default .false. - i.e. don't use data provided by CDEPS inline
```
I think the name of this logical switch is not very descriptive, `use_inline`, inline what?  Maybe `use_cdeps_inline` or something even more specific.

DusanJovic-NOAA commented on cpl/module_inline.F90:
```diff
@@ -0,0 +1,260 @@
+module module_inline
```
Same comment as above, `module_inline` is not very descriptive

uturuncoglu replied:
@DusanJovic-NOAA I agree. Let me try to change them.

DusanJovic-NOAA commented on cpl/module_inline.F90:
```diff
@@ -0,0 +1,260 @@
+      character(len=ESMF_MAXSTR) :: streamfilename, stream_name
+      character(len=ESMF_MAXSTR), allocatable :: file_list(:), var_list(:,:)
+
+      ! skip if use_inline is set to false
+      if (GFS_control%use_inline == .false.) return
```
If `use_inline` is false, there's no point calling this subroutine, only to return immediately. I think it's better to add an if test in the calling routine and skip calling this subroutine altogether.

uturuncoglu replied:
This is mainly protected by `CDEPS_INLINE` in the `fv3/module_fcst_grid_comp.F90`. So, returning in here just allow use change the behavior in run time. If you won't compile code with `CDEPS_INLINE` then there won't be any harm. But, maybe I could put that flag into `fv3/module_fcst_grid_comp.F90` but not sure. Let me know if you want me to do it.

uturuncoglu replied:
Okay. I put them one level up. Removing from here.

DusanJovic-NOAA replied:
That's exactly what I had in mind. When I read the code in `fv3/module_fcst_grid_comp.F90` and I see:

```
#ifdef CDEPS_INLINE
    if (GFS_control%use_cdeps_inline) then
       call cdeps_stream_run(clock, rc)
       if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
    end if
#endif
```
I can reason, okay this code will be executed only if compile option (`CDEPS_INLINE`) is defined, AND if run-time option `use_cdeps_inline` is true. Which I think is better than calling the subroutine always (when `CDEPS_INLINE` is defined), but then checking the run-time option inside the subroutine and returning immediately.

Thanks for making the change.

DusanJovic-NOAA commented on cpl/module_inline.F90:
```diff
@@ -0,0 +1,260 @@
+
+  type(ESMF_Grid) :: grid
+  type(ESMF_Mesh) :: mesh
+  type(ESMF_Field) :: fgrid
```
To me it's a little confusing to name variable of type Field, fgrid. Could this just be named `field`, like mesh is `mesh`, and grid is `grid`

DusanJovic-NOAA replied:
Ok, I now see that `fgrid` here (probably) means field on a grid, since I see local variable `fmesh`, which is probably field on a mesh.

What's the reason for even providing a field on a mesh only to redistributing in to field on a grid? Is there a reason cdeps provides (reads) field on a mesh instead of a field on a grid?

uturuncoglu replied:
@DusanJovic-NOAA For CDEPS, everything is on mesh to make it generic like CMEPS. So, CDEPS is providing field on mesh (even if you are providing data in structured grid) and we need to redist it to the field on grid (in here cubed sphere forecast grid multi tile) and pass it to the CCPP. I hope it makes sense.

DusanJovic-NOAA replied:
Yes, OK. Thanks for explaining.

DusanJovic-NOAA commented on cpl/module_inline.F90:
```diff
@@ -0,0 +1,260 @@
+      real(kind=kp), dimension(:,:), pointer  :: dataptr2d
+
+      ! skip if use_inline is set to false
+      if (GFS_control%use_inline == .false.) return
```
Same comment as above

uturuncoglu committed:
minor changes based on reviwer comments

uturuncoglu committed:
delete file

NickSzapiro-NOAA commented on ccpp/data/GFS_typedefs.F90:
```diff
@@ -3332,6 +3342,19 @@ subroutine coupling_create (Coupling, Model)
     endif
 
+    if (Model%use_cdeps_inline) then
+      allocate (Coupling%tsfco_dat(IM))
+      Coupling%tsfco_dat = clear_val
+      allocate (Coupling%mask_dat(IM))
+      Coupling%mask_dat = clear_val
+      allocate (Coupling%tice_dat(IM))
+      Coupling%tice_dat = clear_val
+      allocate (Coupling%hice_dat(IM))
+      Coupling%hice_dat = clear_val
+      allocate (Coupling%fice_dat(IM))
+      Coupling%fice_dat = clear_val
```
Now or later, is it possible to be more flexible in the number and meaning of these cdeps_inline variables? A different use may want different variables, like https://github.com/ufs-community/ccpp-physics/blob/54672d25027e1c6814ddce8cda5853d824ec1de5/physics/Interstitials/UFS_SCM_NEPTUNE/sfcsub.F#L219-L255

uturuncoglu replied:
@NickSzapiro-NOAA Sure. If we have all the variables in the list, I could change their name. So,

tsfco_dat -> ? not sure since there is no equivalent but maybe sst_dat?
mask_dat -> msk_dat
tice_dat -> ? not sure since there is no equivalent but maybe sit_dat?
hice_dat -> sih_dat
fice_dat -> acn_dat

Let me know if you have any other suggestion.

NickSzapiro-NOAA replied:
Not the variable names themselves. More I'm wondering if this can be an allocatable structure instead (nVariables, variable_dat(IM)) to be more general for these+other variables

Maybe the input netcdf file can be used to specify what the _dat variables mean and what ccpp variables to overwrite in your ccpp-physics/physics/SFC_Models/Data/sfc_data.F

uturuncoglu replied:
@NickSzapiro-NOAA Yes. It is more clear to me. That makes sense. But , this will just make it generic in GFS and CCPP layer. We still need to put those if else statements to `atmos_model` part - `setup_inlinedata` call. Anyway, I could try but it will take little bit time for me to make it due to other commitments. Maybe next week I could look at it.

NickSzapiro-NOAA replied:
Sounds good! We can also make a few PRs to build incrementally as we can

uturuncoglu committed:
fix issue with 32bit build

grantfirl commented:
> @uturuncoglu Maybe you should leave the dimension name in `sfc_data.meta` to be `horizontal_loop_extent`.
> 
> See, for example, how the dimension name of `hflx_lnd` is horizonatal_dimension in GFS_typedefs.meta:
> 
> https://github.com/NOAA-EMC/fv3atm/blob/3145f5b45e244a0949b7082d8bb3057df269b619/ccpp/data/GFS_typedefs.meta#L2992-L2999
> 
> but horizontal_loop_extent in sfc_land.meta:
> 
> https://github.com/ufs-community/ccpp-physics/blob/c86be50a34c600423fcd5df558ff802d5286af10/physics/SFC_Models/Land/sfc_land.meta#L246-L254
> 
> I do not fully understand how this meta specifications are supposed to work, it's all very confusing to me.

Just now seeing this conversation RE: horizontal_loop_extent vs horizontal_dimension. There is some documentation on this that may help clear things up:

- for ccpp-compliant schemes: https://ccpp-techdoc.readthedocs.io/en/latest/CompliantPhysicsParams.html#horizontaldimensionoptionsschemes
- for ccpp host side: https://ccpp-techdoc.readthedocs.io/en/latest/HostSideCoding.html#horizontal-dimension-vs-horizontal-loop-extent

In general, for variables that have horizontal extent, use `horizontal_dimension` in all phases of the scheme EXCEPT run. Use `horizontal_loop_extent` in the run phase of the scheme. On the host side, both must be defined, but the host-side metadata should use `horizontal_dimension`.

It looks like @uturuncoglu and you figured it out by trial and error.

uturuncoglu commented:
@grantfirl Thanks for the extra information. That is really helpful. The confusion was in my end since it was working even with `horizontal_loop_extent` before and failed after sync with the model. Anyway, it is all solved now.

uturuncoglu committed:
convert new data scheme to F90

uturuncoglu committed:
sync model

uturuncoglu committed:
update ccpp/physics

uturuncoglu committed:
update gitmodules file

uturuncoglu committed:
sync model

uturuncoglu committed:
update physics

uturuncoglu committed:
update atmos_cubed_sphere

uturuncoglu committed:
update ccpp physics

uturuncoglu committed:
Merge remote-tracking branch 'origin/develop' into feature/inline_sync

uturuncoglu committed:
sync ccpp physics

uturuncoglu committed:
Merge remote-tracking branch 'upstream/develop' into feature/inline_sync

uturuncoglu committed:
remove missing scheme

uturuncoglu committed:
remove sfc_data from hrrr suite file

grantfirl commented:
@uturuncoglu You need to merge in the latest develop please.

uturuncoglu commented:
@grantfirl let me sync

uturuncoglu committed:
sync ccpp physics

uturuncoglu committed:
Merge remote-tracking branch 'upstream/develop' into feature/inline_sync

uturuncoglu committed:
sync ccpp physics

gspetro-NOAA commented:
Could we get reviews on this sub-PR so that we can process WM parent PR 2807?

gspetro-NOAA commented:
@NickSzapiro-NOAA Could you review this PR (or tag someone who would be better if you're not the right person)?

BrianCurtis-NOAA commented:
ccpp-physics now at https://github.com/ufs-community/ccpp-physics/commit/55886ae3c2c8c95d4e14a72a7115cfe88678178f. Please update hash and revert .gitmodules. Once this is completed it can be merged.

grantfirl commented:
@uturuncoglu The ccpp-physics PR has been merged. Please revert .gitmodules and update the ccpp-physics hash to https://github.com/ufs-community/ccpp-physics/commit/55886ae3c2c8c95d4e14a72a7115cfe88678178f

uturuncoglu committed:
update ccpp physics hash and revert .gitmodules

uturuncoglu commented:
@BrianCurtis-NOAA @grantfirl Okay. I updated UFSATM and point official CCPP physics repo and also point hash.

BrianCurtis-NOAA commented:
looks good, please merge when ready.
